pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -1280,12 +1279,15 @@ private void failPendingMessages(ClientCnx cnx, PulsarClientException ex) {
                 ReferenceCountUtil.safeRelease(op.cmd);
                 op.recycle();
             });
-            semaphore.release(releaseCount.get());
+
             pendingMessages.clear();
             pendingCallbacks.clear();
             if (isBatchMessagingEnabled()) {
                 failPendingBatchMessages(ex);
             }
+
+            semaphore.drainPermits();
+            semaphore.release(conf.getMaxPendingMessages());
```

Comment:
Do you want to ‘clear’ or ‘reset’ the semaphore here? I think we can't ‘clear’ the semaphore here, because acquire semaphore don't get the producer instance lock,  if others is already acquired, here will ‘release’ it and it also might lead the dead lock right?

Reply:
OK,I will fix it